File: src/agent-src/templates/scripts/telemetry/boundary.py
"""Boundary detection + concurrent-safe recording (Phase 2).

Two responsibilities, one module:

1. ``BoundarySession`` — in-process coalescing. Multiple ``add_*`` calls
   within one task / phase-step / tool-call boundary merge into a
   single emitted event (set-union on ``consulted`` / ``applied``).
   Idempotent: calling ``flush()`` twice without new additions is a
   no-op; calling ``add_consulted("skills", ["x"])`` twice records
   ``"x"`` once.

2. ``record_event`` — cross-process durability. Uses ``fcntl.flock``
   (POSIX) so concurrent writers from separate ``./agent-config
   telemetry:record`` invocations cannot interleave inside one JSONL
   line. On non-POSIX (no ``fcntl``) we fall back to a best-effort
   append; the package only ships on POSIX-compatible CI today.

The CLI in ``cli.py`` is the only caller that should touch the log
path directly. Agent-side flows wire through ``BoundarySession``.
"""
from __future__ import annotations

import os
from contextlib import contextmanager
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable, Iterator

try:  # POSIX advisory file locking
    import fcntl  # type: ignore[import-not-found]
    _HAS_FCNTL = True
except ImportError:  # pragma: no cover — Windows / sandbox
    _HAS_FCNTL = False

from .engagement import (
    ALLOWED_BOUNDARY_KINDS,
    ALLOWED_KINDS,
    EngagementEvent,
    EngagementSchemaError,
    now_utc_iso,
)


@dataclass
class BoundarySession:
    """Collect artefact engagements for one boundary, flush once.

    Use as a context manager — ``__exit__`` flushes on clean exit and
    suppresses on exception (so failed tasks don't pollute the log).
    """

    task_id: str
    boundary_kind: str
    log_path: Path
    consulted: dict[str, set[str]] = field(default_factory=dict)
    applied: dict[str, set[str]] = field(default_factory=dict)
    _flushed: bool = False
    _has_data: bool = False

    def __post_init__(self) -> None:
        if self.boundary_kind not in ALLOWED_BOUNDARY_KINDS:
            raise EngagementSchemaError(
                f"boundary_kind must be one of {ALLOWED_BOUNDARY_KINDS!r}"
            )
        if not isinstance(self.task_id, str) or not self.task_id:
            raise EngagementSchemaError("task_id must be a non-empty string")

    def add_consulted(self, kind: str, ids: Iterable[str]) -> None:
        self._merge(self.consulted, kind, ids)

    def add_applied(self, kind: str, ids: Iterable[str]) -> None:
        self._merge(self.applied, kind, ids)

    def _merge(self, bucket: dict[str, set[str]], kind: str, ids: Iterable[str]) -> None:
        if kind not in ALLOWED_KINDS:
            raise EngagementSchemaError(
                f"{kind!r} is not an allowed artefact kind "
                f"(allowed: {ALLOWED_KINDS!r})"
            )
        target = bucket.setdefault(kind, set())
        for art_id in ids:
            if not isinstance(art_id, str) or not art_id:
                raise EngagementSchemaError(
                    f"{kind} ids must be non-empty strings"
                )
            target.add(art_id)
            self._has_data = True

    def to_event(self) -> EngagementEvent:
        return EngagementEvent(
            ts=now_utc_iso(),
            task_id=self.task_id,
            boundary_kind=self.boundary_kind,
            consulted={k: sorted(v) for k, v in self.consulted.items() if v},
            applied={k: sorted(v) for k, v in self.applied.items() if v},
        )

    def flush(self) -> bool:
        """Write one merged event to the log. Returns True if written.

        No-op when already flushed or no data was added — keeps the
        boundary idempotent.
        """
        if self._flushed or not self._has_data:
            return False
        record_event(self.log_path, self.to_event())
        self._flushed = True
        return True

    def __enter__(self) -> "BoundarySession":
        return self

    def __exit__(self, exc_type: object, exc: object, tb: object) -> None:
        if exc_type is None:
            self.flush()
        # On exception: do nothing — failed boundary, no record.


def record_event(log_path: Path, event: EngagementEvent) -> None:
    """Append one event under an exclusive file lock.

    The lock guarantees that two concurrent writers append two
    complete, well-formed lines instead of one interleaved line. We
    open with ``"a"`` so each write atomically extends EOF on POSIX
    once the lock is held.
    """
    event.validate()
    payload = event.to_jsonl().encode("utf-8")
    log_path.parent.mkdir(parents=True, exist_ok=True)
    fd = os.open(log_path, os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o644)
    try:
        if _HAS_FCNTL:
            fcntl.flock(fd, fcntl.LOCK_EX)
        try:
            written = 0
            while written < len(payload):
                written += os.write(fd, payload[written:])
            os.fsync(fd)
        finally:
            if _HAS_FCNTL:
                fcntl.flock(fd, fcntl.LOCK_UN)
    finally:
        os.close(fd)


@contextmanager
def open_boundary(
    task_id: str,
    boundary_kind: str,
    log_path: Path,
) -> Iterator[BoundarySession]:
    """Convenience context manager around ``BoundarySession``.

    >>> with open_boundary("ticket-1", "task", Path(".agent-engagement.jsonl")) as s:
    ...     s.add_consulted("skills", ["php-coder"])
    ...     s.add_applied("skills", ["php-coder"])
    """
    session = BoundarySession(
        task_id=task_id,
        boundary_kind=boundary_kind,
        log_path=log_path,
    )
    with session:
        yield session


__all__ = [
    "BoundarySession",
    "open_boundary",
    "record_event",
]

File: src/agent-src/templates/scripts/telemetry/boundary.ts
/**
 * Boundary detection + concurrent-safe recording (Phase 2).
 *
 * TypeScript twin of `boundary.py` (ADR-096). Byte-for-byte parity on the
 * written JSONL and on the `BoundarySession` coalescing semantics.
 *
 * 1. `BoundarySession` — in-process coalescing. Multiple `add_*` calls
 *    within one boundary merge into a single emitted event (set-union on
 *    `consulted` / `applied`). Idempotent flush.
 * 2. `record_event` — durable append. The Python original takes a POSIX
 *    `fcntl.flock` when available and falls back to a best-effort append
 *    otherwise (`_HAS_FCNTL = False`). Node has no portable advisory lock
 *    in stdlib, so this twin is the documented best-effort-append branch:
 *    the bytes written are identical; only the cross-process interleave
 *    guarantee differs (the package ships on POSIX CI, single-writer in
 *    tests, so output parity is unaffected).
 */
import * as fs from 'node:fs';
import * as path from 'node:path';

import {
    ALLOWED_BOUNDARY_KINDS,
    ALLOWED_KINDS,
    EngagementEvent,
    EngagementSchemaError,
    now_utc_iso,
} from './engagement.js';

function _reprTuple(items: readonly string[]): string {
    const inner = items.map((s) => `'${s}'`).join(', ');
    return items.length === 1 ? `(${inner},)` : `(${inner})`;
}

function _reprScalar(v: unknown): string {
    if (typeof v === 'string') {
        return `'${v}'`;
    }
    return String(v);
}

/** Python `sorted()` on a list of str — lexicographic by code unit. */
function _sortedStrs(arr: Iterable<string>): string[] {
    return [...arr].sort((a, b) => (a < b ? -1 : (a > b ? 1 : 0)));
}

export class BoundarySession {
    task_id: string;
    boundary_kind: string;
    log_path: string;
    consulted: Map<string, Set<string>>;
    applied: Map<string, Set<string>>;
    private _flushed = false;
    private _has_data = false;

    constructor(init: { task_id: string; boundary_kind: string; log_path: string }) {
        this.task_id = init.task_id;
        this.boundary_kind = init.boundary_kind;
        this.log_path = init.log_path;
        this.consulted = new Map();
        this.applied = new Map();
        // __post_init__
        if (!(ALLOWED_BOUNDARY_KINDS as readonly string[]).includes(this.boundary_kind)) {
            throw new EngagementSchemaError(
                `boundary_kind must be one of ${_reprTuple(ALLOWED_BOUNDARY_KINDS)}`,
            );
        }
        if (typeof this.task_id !== 'string' || !this.task_id) {
            throw new EngagementSchemaError('task_id must be a non-empty string');
        }
    }

    add_consulted(kind: string, ids: Iterable<string>): void {
        this._merge(this.consulted, kind, ids);
    }

    add_applied(kind: string, ids: Iterable<string>): void {
        this._merge(this.applied, kind, ids);
    }

    private _merge(bucket: Map<string, Set<string>>, kind: string, ids: Iterable<string>): void {
        if (!(ALLOWED_KINDS as readonly string[]).includes(kind)) {
            throw new EngagementSchemaError(
                `${_reprScalar(kind)} is not an allowed artefact kind `
                + `(allowed: ${_reprTuple(ALLOWED_KINDS)})`,
            );
        }
        let target = bucket.get(kind);
        if (target === undefined) {
            target = new Set<string>();
            bucket.set(kind, target);
        }
        for (const art_id of ids) {
            if (typeof art_id !== 'string' || !art_id) {
                throw new EngagementSchemaError(`${kind} ids must be non-empty strings`);
            }
            target.add(art_id);
            this._has_data = true;
        }
    }

    to_event(): EngagementEvent {
        const consulted: Record<string, string[]> = {};
        for (const [k, v] of this.consulted) {
            if (v.size > 0) {
                consulted[k] = _sortedStrs(v);
            }
        }
        const applied: Record<string, string[]> = {};
        for (const [k, v] of this.applied) {
            if (v.size > 0) {
                applied[k] = _sortedStrs(v);
            }
        }
        return new EngagementEvent({
            ts: now_utc_iso(),
            task_id: this.task_id,
            boundary_kind: this.boundary_kind,
            consulted,
            applied,
        });
    }

    /** Write one merged event to the log. Returns true if written. */
    flush(): boolean {
        if (this._flushed || !this._has_data) {
            return false;
        }
        record_event(this.log_path, this.to_event());
        this._flushed = true;
        return true;
    }
}

/** Append one event. See module docstring for the lock-vs-bytes contract. */
export function record_event(log_path: string, event: EngagementEvent): void {
    event.validate();
    const payload = event.to_jsonl();
    fs.mkdirSync(path.dirname(log_path) || '.', { recursive: true });
    // O_WRONLY | O_CREAT | O_APPEND, 0o644 — best-effort append (no fcntl).
    const fd = fs.openSync(log_path, 'a', 0o644);
    try {
        fs.writeSync(fd, Buffer.from(payload, 'utf-8'));
        fs.fsyncSync(fd);
    } finally {
        fs.closeSync(fd);
    }
}

/**
 * Convenience helper around `BoundarySession`. The Python `open_boundary`
 * is a context manager that flushes on clean exit and suppresses on
 * exception. The TS analogue takes a body callback for the same shape.
 */
export function open_boundary<T>(
    task_id: string,
    boundary_kind: string,
    log_path: string,
    body: (session: BoundarySession) => T,
): T {
    const session = new BoundarySession({ task_id, boundary_kind, log_path });
    const result = body(session);
    session.flush();
    return result;
}
